<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
    <title>Iced: fast and memory efficient ICE normalization &#8212; Iced 0.5.0-git documentation</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet"
        href="../_static/css/bootstrap.min.css"
	media="screen" />
    <script src="../_static/js/jquery.js"
          type="text/javascript"></script>
  <script src="../_static/js/bootstrap.min.js"
          type="text/javascript"></script>


    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-dataframe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-rendered-html.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport'
      content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<link rel="stylesheet"
        href="../_static/css/jhepc.css"
         media="screen" />



  </head><body><div class="navbar">
      <a href="../index.html">
	
      </a>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="../index.html">Home</a></li>
      <li><a href="../install.html">Installation</a></li>
      <li><a href="../documentation.html">Documentation</a></li>
      <li><a href="#">Tutorials</a></li>
      <li><a href="../auto_examples/index.html">Gallery</a></li>
      
    </ul>
  </div>
    <div class="container">
  
  <div class="section" id="iced-tutorial">
<span id="tutorial-menu"></span><h1><a class="toc-backref" href="#id1">Iced tutorial</a><a class="headerlink" href="#iced-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#iced-tutorial" id="id1">Iced tutorial</a></p>
<ul>
<li><p><a class="reference internal" href="#what-is-iced" id="id2">What is iced?</a></p></li>
<li><p><a class="reference internal" href="#working-with-hi-c-data-in-python" id="id3">Working with Hi-C data in Python</a></p></li>
<li><p><a class="reference internal" href="#loading-an-example-dataset" id="id4">Loading an example dataset</a></p></li>
<li><p><a class="reference internal" href="#normalizing-a-data-set-with-ice" id="id5">Normalizing a data set with ICE</a></p></li>
<li><p><a class="reference internal" href="#how-about-cancer-data-sets-loic-and-caic" id="id6">How about cancer data sets? LOIC and CAIC</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Doctest Mode</strong></p>
<blockquote>
<div><p>The code-examples in the above tutorials are written in a
<em>python-console</em> format. If you wish to easily execute these examples
in <strong>IPython</strong>, use:</p>
</div></blockquote>
<p>%doctest_mode</p>
<blockquote>
<div><p>in the IPython-console. You can then simply copy and paste the examples
directly into IPython without having to worry about removing the <strong>&gt;&gt;&gt;</strong>
manually.</p>
</div></blockquote>
</div>
<div class="section" id="what-is-iced">
<h2><a class="toc-backref" href="#id2">What is iced?</a><a class="headerlink" href="#what-is-iced" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">iced</span></code> is a python package that contains normalization techniques for Hi-C data.
It is included in the HiC-pro pipeline, that processes data from raw fastq
files to normalized contact maps. Eventually, <code class="docutils literal notranslate"><span class="pre">iced</span></code> grew bigger than just being a
normalization packages, and contains a number of utilities functions that may
be useful if you are analyzing and processing Hi-C data.</p>
<p>If you use <code class="docutils literal notranslate"><span class="pre">iced</span></code>, please cite:</p>
<p>HiC-Pro: An optimized and flexible pipeline for Hi-C processing. Servant N.,
Varoquaux N., Lajoie BR., Viara E., Chen CJ., Vert JP., Dekker J., Heard E.,
Barillot E. Genome Biology 2015, 16:259 doi:10.1186/s13059-015-0831-x
http://www.genomebiology.com/2015/16/1/259</p>
</div>
<div class="section" id="working-with-hi-c-data-in-python">
<h2><a class="toc-backref" href="#id3">Working with Hi-C data in Python</a><a class="headerlink" href="#working-with-hi-c-data-in-python" title="Permalink to this headline">¶</a></h2>
<p>Hi-C data boils down to a matrix of contact counts. Each row and columns
corresponds to a genomic window, and each entry to the number of times these
genomic windows have been observed to interact with one another. Python
happens to be an excellent language to manipulate matrices, and <code class="docutils literal notranslate"><span class="pre">iced</span></code>
leverages a number of scientific packages that provides nice and easy-to-use
matrix operation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are not familiar with numpy and python, we strongly encourage to
follow the short tutorial of the <a class="reference external" href="http://www.scipy-lectures.org/">scipy lecture notes</a></p>
</div>
</div>
<div class="section" id="loading-an-example-dataset">
<h2><a class="toc-backref" href="#id4">Loading an example dataset</a><a class="headerlink" href="#loading-an-example-dataset" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">iced</span></code> comes with a sample data set that allows you to play a bit with the
package. The sample data set included corresponds to the first 5 chromosomes
of the budding yeast <em>S. cerevisiae</em>. In the following, we start a Python or
IPython interpreter from our shell and load this data set. Our notational
convention is that <code class="docutils literal notranslate"><span class="pre">$</span></code> denotes the shell prompt while <code class="docutils literal notranslate"><span class="pre">&gt;&gt;&gt;</span></code> denotes the
Python interpreter prompt:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ python
&gt;&gt;&gt; from iced import datasets
&gt;&gt;&gt; counts, lengths = datasets.load_sample_yeast()
</pre></div>
</div>
<p>A data set in <code class="docutils literal notranslate"><span class="pre">iced</span></code> is composed of an N by N numpy.ndarray <code class="docutils literal notranslate"><span class="pre">counts</span></code> and a
vector of <code class="docutils literal notranslate"><span class="pre">lengths</span></code> that contains the number of bins per chromosomes. For
our sample data, the vector <code class="docutils literal notranslate"><span class="pre">lengths</span></code> is an ndarray of length 5, underlying
we have here 5 chromosomes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">))</span>
<span class="go">5</span>
</pre></div>
</div>
<p>The contact map <code class="docutils literal notranslate"><span class="pre">counts</span></code> should be squared and symmetric. The shape should
also match the lengths vector:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">(350, 350)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lengths</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
<span class="go">350</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">counts</span></code> matrix is here of size 350 by 350.</p>
<p>You’ve successfully loaded your first Hi-C data!
The corresponding image is the following.</p>
<a class="reference external image-reference" href="../../auto_examples/datasets/plot_yeast_sample.html"><img alt="auto_examples/datasets/images/sphx_glr_plot_yeast_sample_001.png" class="align-center" src="auto_examples/datasets/images/sphx_glr_plot_yeast_sample_001.png" /></a>
</div>
<div class="section" id="normalizing-a-data-set-with-ice">
<h2><a class="toc-backref" href="#id5">Normalizing a data set with ICE</a><a class="headerlink" href="#normalizing-a-data-set-with-ice" title="Permalink to this headline">¶</a></h2>
<p>Now that we have some data loaded, let’s proceed to normalizing it. There are
two normalization algorithms implemented in <cite>iced</cite>: ICE and SCN. ICE is the
most widely used normalization technique on Hi-C data, so this is the one we
will showcase.</p>
<p>ICE is based on a matrix balancing algorithm. The underlying assumptions are
that the contact map suffers from biases that can be decomposable as a product
of regionale biases: <span class="math">C_{ij} = \beta_i \beta_j N_{ij}</span>, where
<span class="math">C_{ij}</span> is the raw contact counts between loci <span class="math">i</span> and <span class="math">j</span>,
<span class="math">N_{ij}</span> the normalized contact counts, and <span class="math">\beta</span> the bias
vector.</p>
<p>Normalizing the data is as simple as follows</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">iced</span> <span class="kn">import</span> <span class="n">normalization</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">normalization</span><span class="o">.</span><span class="n">ICE_normalization</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
<p>But the estimation of the bias vector can be severely problematic in low
coverage regions. In fact, if the matrix is too sparse, the algorithm may not
converge at all! To avoid this, Imakaev et al recommend filtering out a
certain percentage of rows and columns that interact the least. This has to
be performed prior to applying the normalization algorithm:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">iced</span> <span class="kn">import</span> <span class="nb">filter</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="nb">filter</span><span class="o">.</span><span class="n">filter_low_counts</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">percentage</span><span class="o">=</span><span class="mf">0.04</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">normed</span> <span class="o">=</span> <span class="n">normalization</span><span class="o">.</span><span class="n">ICE_normalization</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference external image-reference" href="../../auto_examples/normalization/plot_ice_normalization.html"><img alt="../_images/sphx_glr_plot_ice_normalization_001.png" class="align-center" src="../_images/sphx_glr_plot_ice_normalization_001.png" style="width: 900.0px; height: 300.0px;" /></a>
</div>
<div class="section" id="how-about-cancer-data-sets-loic-and-caic">
<h2><a class="toc-backref" href="#id6">How about cancer data sets? LOIC and CAIC</a><a class="headerlink" href="#how-about-cancer-data-sets-loic-and-caic" title="Permalink to this headline">¶</a></h2>
<p>The last section discusses normalizing Hi-C data using ICE. This method is not
adapted to normalize cancer data sets: several of the assumptions made do not
hold in the presence of copy number variation.</p>
<p><code class="docutils literal notranslate"><span class="pre">iced</span></code> proposes two methods for normalizing data sets with copy number
variations:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">LOIC</span></code>, to preserve enrichment in interactions due to copy number
variations</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">CAIC</span></code> to remove copy number variation effects in the contact count matrix.</p></li>
</ul>
<p>Two examples are available in the gallery, showcasing how to use <code class="docutils literal notranslate"><span class="pre">LOIC</span></code>
and <code class="docutils literal notranslate"><span class="pre">CAIC</span></code> with <code class="docutils literal notranslate"><span class="pre">ice</span></code>.</p>
</div>
</div>


    </div>
<footer class="footer">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2015-2017, iced contributors.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 3.4.1.<br/>
    </p>
</footer>
  </body>
</html>